### DISCUSSION TITLE
[Feature Request] PlayMode_Course/Both

### DISCUSSION THREAD

**Discussion by DarkBahamut162** [TeamRizu/OutFox#457] [03 - Feature Requests]
### Is your feature request related to a problem?

For example, in Mungyodance 3, there is an Extra Mode, which is pretty much the only course mode it has and is being set as `playmode,oni`
This makes it impossible to play nonstop courses in that mode or both in one at all.

### Describe the solution you'd like

A mode that combines both nonstop and oni with one another.
A differentiation between them should be easy as oni courses have `GAINSECONDS` / `LIVES` and nonstop courses have `TIME`.

### Describe alternatives you've considered

There is no alternative. Not that I know of...

**Comment by Scraticus** [comment]
Another great feature request. Converting to discussion so we can archive and add it to the road map.

**Comment by lomdr** [comment]
It was a consideration that I had when I was thinking out the "To Gameplay Flow" for KBX, for the select mode screen I thought of:
Normal (standard 'per song' gameplay), Extreme mode (Extra Mode from MGD3), Course Mode (DDR-like Nonstop and Oni Courses), and Extreme Course Mode (ITG's Marathon mode).  Of course, this can be added to any other mode that can 'cleanly' support mod charts as well.  Though, IIRC courses haven't really been updated since the 3.X days and there's some conventions with the CRS files that makes it hard for them to share (Exact directories are needed per song or that song doesn't display, the entire course if it has 'no songs')

**Comment by Scraticus** [comment]
Yea with all the new modes we need to revamp the entire hashing system for courses.

I want to remove the filepath method, we should by now be using a hash number. So as long as the simfile/contents do not change the number is all that is required and the game should find it. But it's a few months away yet before this will be implemented.

Courses will be getting work in the spring so feel free to add more ideas and suggestions here of course.

**Comment by DarkBahamut162** [comment]
I was mostly confused about what lomdr said.
But thinking about it, maybe one could do a pre-sort on certain things within the metric:

Sorting Mod/Lua Courses*
-----------------------
- NoModsAndLua
- ModsOnly
- LuaOnly
- ModsAndLuaOnly

*I've already seen the somewhat old stream showing off Lua-Mods during gameplay like ITG's modifiers during course mode. Maybe a "hasLua()" function would be something for that to be added as well. It would made some of my code I did in my edited ITG3Encore theme obsolete by searching for & marking them like "hasMods".

Sorting PlayMode Type (if on PlayMode_Both/Course)
---------------------
- LifeOnly
- TimeOnly
- NonstopOnly
- LifeAndTime (PlayMode_Oni)
- AllModes (default)

Checking for what Course is what Mode:
--------------------------------------
```
if CourseEntry:GetGainLives > 0 then
	PlayMode_Oni (LifeLine)
elseif CourseEntry:GetGainSecond > 0 then
	PlayMode_Oni (Survival)
else
	PlayMode_Nonstop
end
```

From what lomdr said with certain mods, it would look like this:

**Normal** (PlayMode_Regular)
**Extreme Mode*** (PlayMode_Oni;LifesOnly)
**Course Mode** (PlayMode_Both/Course;NoModsAndLua)
**Extreme Course Mode** (PlayMode_Nonstop;ModsAndLuaOnly/ModsOnly)

*as I've already said: MungyoDance's Extra Mode already is PlayMode_Oni, which has both Lifes and Time. If it is meant "LifesOnly" then it would be what I've just said.


Food for thought...


But I'm still a little bit confused about KBX and/or IIRC Courses and what they are. But depending on what they are, it should be possible to sort for them as well (maybe).
Anyway: I'm not sure about "hashes" as I'm not that knowledgeable/far in that.
Course Mode only needs the somewhat same functions like PlayMode-Regular while being potentially sortable.
